source/server/overload_manager_impl.h

```diff
@@ -52,6 +52,53 @@ class OverloadAction {
   Stats::Gauge& scale_percent_gauge_;
 };
 
+// Simple table that converts strings into Symbol instances. Symbols are guaranteed to start at 0
+// and be indexed sequentially.
+class NamedOverloadActionSymbolTable {
+public:
+  class Symbol {
+  public:
+    // Allow copy construction everywhere.
+    Symbol(const Symbol&) = default;
+
+    // Returns the index of the symbol in the table.
+    size_t index() const { return index_; }
+
+    // Support use as a map key.
+    bool operator==(const Symbol other) { return other.index_ == index_; }
+
+    // Support absl::Hash.
+    template <typename H>
+    friend H AbslHashValue(H h, const Symbol& s) { // NOLINT(readability-identifier-naming)
+      return H::combine(std::move(h), s.index_);
+    }
+
+  private:
+    friend class NamedOverloadActionSymbolTable;
+    // Only the symbol table class can create Symbol instances from indices.
+    explicit Symbol(size_t index) : index_(index) {}
+
+    size_t index_;
+  };
+
+  // Finds an existing or adds a new entry for the given name.
+  Symbol get(absl::string_view name);
+
+  // Returns the symbol for the name if there is one, otherwise nullopt.
+  absl::optional<Symbol> lookup(absl::string_view string) const;
+
+  // Translates a symbol back into a name.
+  const absl::string_view name(Symbol symbol) const;
+
+  // Returns the number of symbols in the table. All symbols are guaranteed to have an index less
+  // than size().
+  size_t size() const { return table_.size(); }
+
+private:
+  absl::flat_hash_map<std::string, size_t> table_;
+  std::vector<std::string> names_;
```

Comment:
it'd be a little nicer not to have to have separate copies of the strings for the map and vector.

Reply:
I'm sure there's a way to do it with string_view but it wasn't immediately obvious since the vector can get resized.

Reply:
both the flat_hash_map and the vector can move their keys around, and in modern STL the string data can be inlined into the string object so it moves too, if it's <22 chars long. So it's not obvious to me how string_view can work.

but node_hash_map solves the problem, as does `std::vector<std::unique_ptr<std::string>>`. You could also use a raw nul-terminated `char*` if you cared about storage and could get that past review :)

Reply:
Yep, both of those work, but at the cost of an extra pointer indirection for some operations. If I had to choose, I'd prefer `std::vector<std::unique_ptr<std::string>>` since it's only used for retrieving the actual string, which is only used for logging.